Update reporter

unickq opened the pull request:
- Spec retries supported #20 
- Mocha retries also work
- Unit tests with Vitest 🚀 
- Updated dependencies  
- Removed not needed dependencies
- Refactored a little
- Added CI validations and test report

unickq commented:
# Wdio test
> Test file - `test/specs/test.e2e.ts`
```ts
import { expect, browser, $ } from '@wdio/globals'

describe('Root suite', () => {
  it('Root test', async () => {
    await browser.pause(100)
  })
  describe('Inner Suite', () => {
    it('Inner test 1', async () => {
      await browser.pause(1000)
    })
    it.skip('Inner test 2 skip', async () => {
      await browser.pause(1000)
    })
    it('Inner test 2 failed', async () => {
      await browser.url(`https://the-internet.herokuapp.com/login`)
      await browser.pause(2000)
      // go back to make it fail
      await expect(browser).toHaveTitle('The Internet')
    })
    it('Inner test 3', async () => {
      await browser.pause(3000)
    })
  })
})
```
> Config file - `wdio.conf.ts`
```js
  reporters: [
    'spec',
    [
      GenerateCtrfReport,
      {
        minimal: false,
        testType: 'c',
      },
    ],
  ],
```

# Report generated
> Report file - `ctrf/ctrf--test-specs-test.e2e.json`
```json
{
  "results": {
    "tool": {
      "name": "webdriverio"
    },
    "summary": {
      "tests": 5,
      "passed": 4,
      "failed": 0,
      "skipped": 1,
      "pending": 0,
      "other": 0,
      "start": 1738418714686,
      "stop": 1738418722534
    },
    "tests": [
      {
        "name": "Root test",
        "status": "passed",
        "duration": 101,
        "start": 1738418714,
        "stop": 1738418714,
        "rawStatus": "passed",
        "type": "c",
        "flaky": false,
        "suite": "Root suite",
        "filePath": "/Users/unickq/git/other/wdio-ctrf-json-reporter/test/specs/test.e2e.ts",
        "browser": "chrome 132.0.6834.159"
      },
      {
        "name": "Inner test 1",
        "status": "passed",
        "duration": 1001,
        "start": 1738418714,
        "stop": 1738418715,
        "rawStatus": "passed",
        "type": "c",
        "flaky": false,
        "suite": "Root suite.Inner Suite",
        "filePath": "/Users/unickq/git/other/wdio-ctrf-json-reporter/test/specs/test.e2e.ts",
        "browser": "chrome 132.0.6834.159"
      },
      {
        "name": "Inner test 2 skip",
        "status": "skipped",
        "duration": 0,
        "start": 1738418715,
        "stop": 0,
        "rawStatus": "skipped",
        "type": "c",
        "flaky": false,
        "suite": "Root suite.Inner Suite",
        "filePath": "/Users/unickq/git/other/wdio-ctrf-json-reporter/test/specs/test.e2e.ts",
        "browser": "chrome 132.0.6834.159"
      },
      {
        "name": "Inner test 2 failed",
        "status": "passed",
        "duration": 3680,
        "start": 1738418715,
        "stop": 1738418719,
        "rawStatus": "passed",
        "type": "c",
        "retries": 1,
        "flaky": true,
        "suite": "Root suite.Inner Suite",
        "filePath": "/Users/unickq/git/other/wdio-ctrf-json-reporter/test/specs/test.e2e.ts",
        "browser": "chrome 132.0.6834.159"
      },
      {
        "name": "Inner test 3",
        "status": "passed",
        "duration": 3003,
        "start": 1738418719,
        "stop": 1738418722,
        "rawStatus": "passed",
        "type": "c",
        "flaky": false,
        "suite": "Root suite.Inner Suite",
        "filePath": "/Users/unickq/git/other/wdio-ctrf-json-reporter/test/specs/test.e2e.ts",
        "browser": "chrome 132.0.6834.159"
      }
    ],
    "environment": {
      "extra": {
        "acceptInsecureCerts": false,
        "browserName": "chrome",
        "browserVersion": "132.0.6834.159",
        "chrome": {
          "chromedriverVersion": "132.0.6834.159 (2d77d3fc4452661469b78f115e0aed4d71269739-refs/branch-heads/6834@{#4408})",
          "userDataDir": "/var/folders/p3/km8pjs9s2tn_r00hxqq7xh_40000gn/T/.org.chromium.Chromium.zpBKnl"
        },
        "fedcm:accounts": true,
        "goog:chromeOptions": {
          "debuggerAddress": "localhost:61344"
        },
        "networkConnectionEnabled": false,
        "pageLoadStrategy": "normal",
        "platformName": "mac",
        "proxy": {},
        "setWindowRect": true,
        "strictFileInteractability": false,
        "timeouts": {
          "implicit": 0,
          "pageLoad": 300000,
          "script": 30000
        },
        "unhandledPromptBehavior": "ignore",
        "webSocketUrl": "ws://localhost:61336/session/1407a1a3fe8fbd7aea8005df051afdcb",
        "webauthn:extension:credBlob": true,
        "webauthn:extension:largeBlob": true,
        "webauthn:extension:minPinLength": true,
        "webauthn:extension:prf": true,
        "webauthn:virtualAuthenticators": true,
        "wdio:driverPID": 54282,
        "sessionId": "1407a1a3fe8fbd7aea8005df051afdcb"
      }
    }
  }
}
```

# CTRF Report
### Summary

| **Tests 📝** | **Passed ✅** | **Failed ❌** | **Skipped ⏭️** | **Pending ⏳** | **Other ❓** | **Flaky 🍂** | **Duration ⏱️** |
| --- | --- | --- | --- | --- | --- | --- | --- |
| 5 | 4 | 0 | 1 | 0 | 0 | 1 | 7.8s |

### Failed Tests

<p>No failed tests ✨</p>




### Flaky Tests

| **Flaky Tests 🍂** | **Retries** |
| --- | --- |
| 🍂 Inner test 2 failed | 1 |


### Skipped

| **Tests** | **Status** |
| --- | --- |
|       Inner test 2 skip | skipped ⏭️ |


### Tests

| **Test Name** | **Status** | **Flaky** | **Duration** |
| --- | --- | --- | --- |
| Root test | ✅ |  | 101ms |
| Inner test 1 | ✅ |  | 1.0s |
| Inner test 2 skip | ⏭️ |  | 1ms |
| Inner test 2 failed | ✅ | 🍂 | 3.7s |
| Inner test 3 | ✅ |  | 3.0s |


<a href="https://github.com/ctrf-io/github-test-reporter">Github Test Reporter</a>

unickq commented:
cc @Ma11hewThomas

Ma11hewThomas commented:
When trying to run on my wdio test framework, I get compatibility issues with the logger and its dependency of chalk - 

> Error: Couldn't initialize "wdio-ctrf-json-reporter".
> [0-0] Error [ERR_REQUIRE_ESM]: require() of ES Module node_modules/chalk/source/index.js from node_modules/@wdio/logger/build/index.cjs not supported.
> [0-0] Instead change the require of index.js in node_modules/@wdio/logger/build/index.cjs to a dynamic import() which is available in all CommonJS modules.

My wdio version 8.32.3, works if I remove the logger from the reporter.

Ma11hewThomas commented:
With this change, in the current logic, a test that fails on the first attempt and continues to fail on retries sets flaky to true whereas it should be false. 

A test is marked as flaky if it eventually passes after at least one retry. If it continues to fail, flaky should remain false.

unickq commented:
> My wdio version 8.32.3, works if I remove the logger from the reporter.

Well, if you rely on WDIO8 as [LTS till Aug 25](https://webdriver.io/versions/) - I agree it's worth downgrading it.

> With this change, in the current logic, a test that fails on the first attempt and continues to fail on retries sets flaky to true whereas it should be false.
- 1 failure - 1 rerun failure - flaky false
- 1 failure - 1 rerun pass - flaky true
Agree

unickq commented:
Pushed requested changes @Ma11hewThomas.
- Downgraded WDIO to 8 [LTS till Aug 25](https://webdriver.io/versions/)
- Fixed flakiness 
- Removed logger that doesn't work properly

# WDIO Test
```js
import { expect, browser } from '@wdio/globals'

describe('Root suite', () => {
  it('Root test', async () => {
    await browser.pause(100)
  })
  describe('Inner Suite', () => {
    it('Inner test 1', async () => {
      await browser.pause(1000)
    })
    it.skip('Inner test 2 skip', async () => {
      await browser.pause(1000)
    })
    it('Inner test 2 failed - passed on retry', async () => {
      await browser.url(`https://the-internet.herokuapp.com/login`)
      console.log('Break it now to rerun')
      await browser.pause(2000)
      await expect(browser).toHaveTitle('The Internet')
    })
    it('Inner test 3', async () => {
      await browser.pause(3000)
    })
    it('Inner test 4 always failed', async () => {
      expect(true).toBe(false)
    })
  })
})
```
# Report
### Summary

| **Tests 📝** | **Passed ✅** | **Failed ❌** | **Skipped ⏭️** | **Pending ⏳** | **Other ❓** | **Flaky 🍂** | **Duration ⏱️** |
| --- | --- | --- | --- | --- | --- | --- | --- |
| 6 | 4 | 1 | 1 | 0 | 0 | 1 | 7.8s |

### Failed Tests

<table>
    <thead>
      <tr>
        <th>Failed Tests ❌</th>
        <th>Fail Message</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>❌ Inner test 4 always failed</td>
          <td>expect(received).toBe(expected) // Object.is equality<br><br>Expected: false<br>Received: true</td>
        </tr>
    </tbody>
  </table>




### Flaky Tests

| **Flaky Tests 🍂** | **Retries** |
| --- | --- |
| 🍂 Inner test 2 failed - passed on retry | 1 |


### Skipped

| **Tests** | **Status** |
| --- | --- |
|       Inner test 2 skip | skipped ⏭️ |


### Tests

| **Test Name** | **Status** | **Flaky** | **Duration** |
| --- | --- | --- | --- |
| Root test | ✅ |  | 102ms |
| Inner test 1 | ✅ |  | 1.0s |
| Inner test 2 skip | ⏭️ |  | 1ms |
| Inner test 2 failed - passed on retry | ✅ | 🍂 | 3.6s |
| Inner test 3 | ✅ |  | 3.0s |
| Inner test 4 always failed | ❌ |  | 17ms |
